x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -0,0 +1,349 @@
+            assert XPackUser.INSTANCE.roles().length == 1;
+            roleActionListener.onResponse(XPackUser.ROLE);
+            return;
+        }
+        if (XPackSecurityUser.is(user)) {
+            roleActionListener.onResponse(ReservedRolesStore.SUPERUSER_ROLE);
+            return;
+        }
+
+        Set<String> roleNames = new HashSet<>(Arrays.asList(user.roles()));
+        if (isAnonymousEnabled && anonymousUser.equals(user) == false) {
+            if (anonymousUser.roles().length == 0) {
+                throw new IllegalStateException("anonymous is only enabled when the anonymous user has roles");
+            }
+            Collections.addAll(roleNames, anonymousUser.roles());
+        }
+
+        if (roleNames.isEmpty()) {
+            roleActionListener.onResponse(Role.EMPTY);
+        } else if (roleNames.contains(ReservedRolesStore.SUPERUSER_ROLE_DESCRIPTOR.getName())) {
+            roleActionListener.onResponse(ReservedRolesStore.SUPERUSER_ROLE);
+        } else {
+            rolesStore.roles(roleNames, fieldPermissionsCache, roleActionListener);
+        }
+    }
+
+    @Override
+    public void authorizeRunAs(Authentication authentication, TransportRequest request, String action, AuthorizationInfo authorizationInfo,
+                               ActionListener<AuthorizationResult> listener) {
+        if (authorizationInfo instanceof RBACAuthorizationInfo) {
+            final Role role = ((RBACAuthorizationInfo) authorizationInfo).getAuthenticatedUserAuthorizationInfo().getRole();
+            listener.onResponse(new AuthorizationResult(role.runAs().check(authentication.getUser().principal())));
+        } else {
+            listener.onFailure(new IllegalArgumentException("unsupported authorization info:" +
+                authorizationInfo.getClass().getSimpleName()));
+        }
+    }
+
+    @Override
+    public void authorizeClusterAction(Authentication authentication, TransportRequest request, String action,
+                                       AuthorizationInfo authorizationInfo, ActionListener<AuthorizationResult> listener) {
+        if (authorizationInfo instanceof RBACAuthorizationInfo) {
+            final Role role = ((RBACAuthorizationInfo) authorizationInfo).getRole();
+            if (role.cluster().check(action, request)) {
+                listener.onResponse(AuthorizationResult.granted());
+            } else if (checkSameUserPermissions(action, request, authentication)) {
+                listener.onResponse(AuthorizationResult.granted());
+            } else {
+                listener.onResponse(AuthorizationResult.deny());
+            }
+        } else {
+            listener.onFailure(new IllegalArgumentException("unsupported authorization info:" +
+                authorizationInfo.getClass().getSimpleName()));
+        }
+    }
+
+    @Override
+    public boolean checkSameUserPermissions(String action, TransportRequest request, Authentication authentication) {
```

Comment:
I think we should call this in the `AuthorizationService` and not inside the `RBACEngine#authorizeClusterAction`